Candle Kernels tries to build wmma kernels on unsupported hardware.

Even when the CUDA_COMPUTE_CAP is set ala https://github.com/huggingface/candle?tab=readme-ov-file#docker-build

This is a problem because I can't have candle with cargo in any GitHub actions, even if they are on the GPU because the T4 runners are so old. 
Is there some way that Candle-Kernels could have feature flags or something to gate these kernels so it can build on older hardware? 

```src/moe/moe_wmma.cu(110): warning #177-D: variable "laneId" was declared but never referenced
        const int laneId = threadId % 32;
                  ^
  Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"
  src/moe/moe_wmma.cu(172): error: incomplete type "nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, nv_bfloat16, nvcuda::wmma::row_major>" (aka "nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, __nv_bfloat16, nvcuda::wmma::row_major>") is not allowed
                fragment<matrix_a, WMMA_M, WMMA_N, WMMA_K, T, row_major> a_frag;
                                                                         ^
            detected during instantiation of "void vllm_rs::moe_gemm_grouped_kernel<T,WMMA_M,WMMA_N,WARPS_N>(const T *, const T *, const int32_t *, const int32_t *, const float *, T *, int, int, int32_t, int32_t, int32_t) [with T=nv_bfloat16, WMMA_M=16, WMMA_N=16, WARPS_N=2]" at line 279
  src/moe/moe_wmma.cu(173): error: incomplete type "nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, nv_bfloat16, nvcuda::wmma::col_major>" (aka "nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, __nv_bfloat16, nvcuda::wmma::col_major>") is not allowed
                fragment<matrix_b, WMMA_M, WMMA_N, WMMA_K, T, col_major> b_frag;
                                                                         ^
            detected during instantiation of "void vllm_rs::moe_gemm_grouped_kernel<T,WMMA_M,WMMA_N,WARPS_N>(const T *, const T *, const int32_t *, const int32_t *, const float *, T *, int, int, int32_t, int32_t, int32_t) [with T=nv_bfloat16, WMMA_M=16, WMMA_N=16, WARPS_N=2]" at line 279
  src/moe/moe_wmma.cu(172): error: incomplete type "nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 8, 32, 16, nv_bfloat16, nvcuda::wmma::row_major>" (aka "nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 8, 32, 16, __nv_bfloat16, nvcuda::wmma::row_major>") is not allowed
                fragment<matrix_a, WMMA_M, WMMA_N, WMMA_K, T, row_major> a_frag;
                                                                         ^
            detected during instantiation of "void vllm_rs::moe_gemm_grouped_kernel<T,WMMA_M,WMMA_N,WARPS_N>(const T *, const T *, const int32_t *, const int32_t *, const float *, T *, int, int, int32_t, int32_t, int32_t) [with T=nv_bfloat16, WMMA_M=8, WMMA_N=32, WARPS_N=1]" at line 281
  src/moe/moe_wmma.cu(173): error: incomplete type "nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 8, 32, 16, nv_bfloat16, nvcuda::wmma::col_major>" (aka "nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 8, 32, 16, __nv_bfloat16, nvcuda::wmma::col_major>") is not allowed
                fragment<matrix_b, WMMA_M, WMMA_N, WMMA_K, T, col_major> b_frag;
                                                                         ^
            detected during instantiation of "void vllm_rs::moe_gemm_grouped_kernel<T,WMMA_M,WMMA_N,WARPS_N>(const T *, const T *, const int32_t *, const int32_t *, const float *, T *, int, int, int32_t, int32_t, int32_t) [with T=nv_bfloat16, WMMA_M=8, WMMA_N=32, WARPS_N=1]" at line 281
  src/moe/moe_wmma.cu(37): warning #177-D: variable "vllm_rs::NUM_VECS" was declared but never referenced
    constexpr int NUM_VECS = 32;
                    fragment<matrix_b, WMMA_M, WMMA_N, WMMA_K, T, col_major> b_frag;
                                                                             ^
            detected during instantiation of "void moe_gemm_gguf_prefill_kernel<T,qk,block_q_t,wrap_size>(const T *, const uint8_t *, const int32_t *, const int32_t *, const float *, float *, int, int, int32_t, int32_t, int32_t, int) [with T=nv_bfloat16, qk=256, block_q_t=block_q6_K, wrap_size=64]" at line 417
  12 errors detected in the compilation of "src/moe/moe_wmma_gguf.cu".
  src/quantized.cu(4405): warning #177-D: variable "ncols_y" was declared but never referenced
        constexpr int ncols_y = 1;
                      ^
  Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"
  src/quantized.cu(4417): warning #177-D: variable "blocks_per_col_y" was declared but never referenced
        const int blocks_per_col_y = k_padded / 32;
                  ^
  thread 'main' (3560) panicked at /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/bindgen_cuda-0.1.6/src/lib.rs:405:13:
  nvcc error while compiling "src/moe/moe_wmma.cu":
  # CLI "nvcc" "--gpu-architecture=sm_75" "--ptx" "--default-stream" "per-thread" "--output-directory" "/__w/luminal/luminal/target/debug/build/candle-kernels-edc31d50e350577b/out" "--expt-relaxed-constexpr" "-std=c++17" "-O3" "-Isrc" "-Isrc/moe" "-I/usr/local/cuda/include" "src/moe/moe_wmma.cu" 
  # stdout
  # stderr
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
Error: Process completed with exit code 101.```

--- a/candle-kernels/build.rs
+++ b/candle-kernels/build.rs
@@ -31,6 +31,15 @@ fn main() -> Result<()> {
         .arg("-std=c++17")
         .arg("-O3");
 
+    // Disable bf16 WMMA kernels on GPUs older than sm_80 (Ampere).
+    // bf16 WMMA fragments require compute capability >= 8.0.
+    let compute_cap = cudaforge::detect_compute_cap()
+        .map(|arch| arch.base())
+        .unwrap_or(80);
+    if compute_cap < 80 {
+        moe_builder = moe_builder.arg("-DNO_BF16_KERNEL");
+    }
+
     let mut is_target_msvc = false;
     if let Ok(target) = std::env::var("TARGET") {
         if target.contains("msvc") {

--- a/candle-kernels/src/moe/moe_wmma.cu
+++ b/candle-kernels/src/moe/moe_wmma.cu
@@ -274,11 +274,14 @@ extern "C" void moe_gemm_wmma(
             // we use smaller M_tile and larger N_tile for decoding
             LAUNCH_MOE_WMMA(half, 8, 32, 1)
         }
-    } else if (data_type == 1) { // bfloat16
+    }
+#ifndef NO_BF16_KERNEL
+    else if (data_type == 1) { // bfloat16
         if (is_prefill) {
             LAUNCH_MOE_WMMA(nv_bfloat16, 16, 16, 2)
         } else {
             LAUNCH_MOE_WMMA(nv_bfloat16, 8, 32, 1)
         }
     }
+#endif
 }

